From 2b383181899ccedc12485cddc4c78f7318bdc3c5 Mon Sep 17 00:00:00 2001
From: Rahul Tanwar <rtanwar@maxlinear.com>
Date: Mon, 28 Jun 2021 19:44:47 +0800
Subject: [PATCH] PONRTSYS-9076: Fix 'system hangs after kernel panic'

Remove incorrect polling for reset status of GPHY_POWER_DOWN
in RCU driver's reset_assert routine. GPHY_POWER_DOWN is not
a reset request bit unlike other bits in RST_REQ registers.
So it does not have a corresponding status bit. Add fix to
remove polling for the non existent status bit for it.
---
 drivers/reset/reset-intel-syscon.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/reset/reset-intel-syscon.c b/drivers/reset/reset-intel-syscon.c
index 79ae0fe4ebf0..e53eb0dd20f4 100644
--- a/drivers/reset/reset-intel-syscon.c
+++ b/drivers/reset/reset-intel-syscon.c
@@ -19,6 +19,9 @@
 #define RCU_RST_REQ		0x0048
 #define STR_MAX_LEN		20
 
+#define GPHY_PWR_DOWN_BIT	27
+#define GPHY_PWR_DOWN_ID	((RCU_RST_REQ << 8) | GPHY_PWR_DOWN_BIT)
+
 /* reset platform data */
 #define to_reset_data(x)	container_of(x, struct intel_reset_data, rcdev)
 
@@ -117,6 +120,15 @@ static int intel_assert_device_timeout(struct reset_controller_dev *rcdev,
 		return ret;
 	}
 
+	if (id == GPHY_PWR_DOWN_ID)
+		/*
+		 * GPHY_POWER_DOWN bit is not a reset request bit. So it does
+		 * not have a corresponding reset status bit. This is unlike
+		 * other reset request bits in these registers. No need to check
+		 * reset status for GPHY_POWER_DOWN assertion.
+		 */
+		return 0;
+
 	regstoff = intel_stat_reg_off(data, regoff);
 	return regmap_read_poll_timeout(data->regmap, regstoff, val,
 					!!(val & BIT(regbit)), sleep_us, 200);
@@ -145,6 +157,10 @@ static int intel_deassert_device(struct reset_controller_dev *rcdev,
 			"Failed to set reset deassert bit %d\n", ret);
 		return ret;
 	}
+
+	if (id == GPHY_PWR_DOWN_ID) /* no status bit for this */
+		return 0;
+
 	regstoff = intel_stat_reg_off(data, regoff);
 	return regmap_read_poll_timeout(data->regmap, regstoff, val,
 					!(val & BIT(regbit)), 20, 200);
